File: packages/transport/mcp/go/authenticate.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Framework-neutral MCP authentication: bearer parse, identity verify, revocation check.

package transportmcp

import (
	"errors"
	"strings"

	"github.com/garudex-labs/caracal/identity"
	"github.com/garudex-labs/caracal/revocation"
)

// Options configures the MCP authentication pipeline.
type Options struct {
	Issuer               string
	Audience             string
	ZoneID               string
	RequiredScopes       []string
	RequireAgent         bool
	RequireDelegation    bool
	RequireChainContains []string
	Revocations          revocation.Store
}

// ErrorCode names every transport-neutral failure mode.
type ErrorCode string

const (
	ErrMissingToken      ErrorCode = "missing_token"
	ErrInvalidToken      ErrorCode = "invalid_token"
	ErrInvalidZone       ErrorCode = "invalid_zone"
	ErrInsufficientScope ErrorCode = "insufficient_scope"
	ErrSessionRevoked    ErrorCode = "session_revoked"
)

// AuthError is the typed failure returned by Authenticate.
type AuthError struct {
	Code        ErrorCode
	Description string
}

func (e *AuthError) Error() string { return e.Description }

// ExtractBearer pulls a non-empty bearer token from an Authorization header value, or returns false.
func ExtractBearer(authHeader string) (string, bool) {
	if !strings.HasPrefix(authHeader, "Bearer ") {
		return "", false
	}
	token := strings.TrimSpace(strings.TrimPrefix(authHeader, "Bearer "))
	if token == "" {
		return "", false
	}
	return token, true
}

// Authenticate verifies a token against identity and revocation, returning typed claims or an AuthError.
func Authenticate(token string, opts Options) (identity.Claims, *AuthError) {
	if token == "" {
		return identity.Claims{}, &AuthError{Code: ErrMissingToken, Description: "Missing bearer token"}
	}
	cfg := identity.Config{
		Issuer:               opts.Issuer,
		Audience:             opts.Audience,
		ZoneID:               opts.ZoneID,
		RequiredScopes:       opts.RequiredScopes,
		RequireAgent:         opts.RequireAgent,
		RequireDelegation:    opts.RequireDelegation,
		RequireChainContains: opts.RequireChainContains,
	}
	claims, err := identity.Verify(token, cfg)
	if err != nil {
		var scopeErr *identity.ScopeMissingError
		switch {
		case errors.As(err, &scopeErr):
			return identity.Claims{}, &AuthError{Code: ErrInsufficientScope, Description: "Missing scope: " + scopeErr.Scope}
		case errors.Is(err, identity.ErrZoneInvalid):
			return identity.Claims{}, &AuthError{Code: ErrInvalidZone, Description: "Token zone validation failed"}
		default:
			return identity.Claims{}, &AuthError{Code: ErrInvalidToken, Description: "Token validation failed"}
		}
	}
	if opts.Revocations != nil && claims.Sid != "" && opts.Revocations.IsRevoked(claims.Sid) {
		return identity.Claims{}, &AuthError{Code: ErrSessionRevoked, Description: "Session revoked"}
	}
	return claims, nil
}

File: packages/transport/mcp/ts/src/authenticate.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Transport-neutral MCP authentication: bearer verify, revocation check, typed result.

import {
  AgentIdentityRequiredError,
  ChainMismatchError,
  DelegationRequiredError,
  HopCountExceededError,
  ScopeInsufficientError,
  TokenInvalidError,
  ZoneInvalidError,
  MANDATE_USE_RESOURCE,
  verify,
  type JwtConfig,
} from '@caracalai/identity'
import type { RevocationStore } from '@caracalai/revocation'
import type { AuthError, AuthResult, Principal } from './types.js'

export type AuthDeps = JwtConfig & { revocations: RevocationStore }

const BEARER_SCHEME = 'bearer'

export function extractBearer(authHeader: string | undefined): string | null {
  if (authHeader === undefined || authHeader.slice(0, BEARER_SCHEME.length).toLowerCase() !== BEARER_SCHEME) return null
  const value = authHeader.slice(BEARER_SCHEME.length)
  if (value.length === value.trimStart().length) return null
  const token = value.trim()
  return token === '' ? null : token
}

export async function authenticate(token: string, deps: AuthDeps): Promise<AuthResult> {
  if (!token) {
    return { ok: false, error: { code: 'missing_token', description: 'Missing bearer token' } }
  }

  try {
    const { revocations, ...jwtConfig } = deps
    const claims = await verify(token, { ...jwtConfig, requiredUse: jwtConfig.requiredUse ?? MANDATE_USE_RESOURCE })
    if (!revocations || typeof revocations.isRevoked !== 'function') {
      return { ok: false, error: { code: 'invalid_token', description: 'Revocation store required' } }
    }
    const activeError = await checkActiveAuthority(claims, revocations)
    if (activeError) {
      return { ok: false, error: activeError }
    }
    return { ok: true, principal: claims }
  } catch (err) {
    if (err instanceof ScopeInsufficientError) {
      return { ok: false, error: { code: 'insufficient_scope', description: err.message } }
    }
    if (err instanceof AgentIdentityRequiredError) {
      return { ok: false, error: { code: 'agent_required', description: err.message } }
    }
    if (err instanceof DelegationRequiredError) {
      return { ok: false, error: { code: 'delegation_required', description: err.message } }
    }
    if (err instanceof ChainMismatchError) {
      return { ok: false, error: { code: 'chain_mismatch', description: err.message } }
    }
    if (err instanceof HopCountExceededError) {
      return { ok: false, error: { code: 'hop_count_exceeded', description: err.message } }
    }
    if (err instanceof ZoneInvalidError) {
      return { ok: false, error: { code: 'invalid_zone', description: 'Token zone validation failed' } }
    }
    if (err instanceof TokenInvalidError) {
      return { ok: false, error: { code: 'invalid_token', description: 'Token validation failed' } }
    }
    return { ok: false, error: { code: 'invalid_token', description: 'Token validation failed' } }
  }
}

export async function checkActiveAuthority(claims: Principal, revocations: RevocationStore, nowMs = Date.now()): Promise<AuthError | null> {
  if (!claims.sid) {
    return { code: 'invalid_token', description: 'Token validation failed' }
  }
  if (claims.expiresAt * 1000 <= nowMs) {
    return { code: 'invalid_token', description: 'Token expired during execution' }
  }
  for (const anchor of revocationAnchors(claims)) {
    if (await revocations.isRevoked(anchor)) {
      return { code: 'session_revoked', description: 'Session revoked' }
    }
  }
  return null
}

function revocationAnchors(claims: Principal): string[] {
  const anchors = [
    claims.sid,
    claims.rootSid,
    claims.agentSessionId,
    claims.delegationEdgeId,
  ].filter((value): value is string => typeof value === 'string' && value !== '')
  return [...new Set(anchors)]
}
